### ISSUE TITLE
Unable to get proper results for imbalanced dataset

### ISSUE THREAD

**Issue by Indhujamari** [qubvel/segmentation_models#445]
I am currently working on the multi-class segmentation of Agricultural aerial images. I have about 7 classes which are highly imbalanced both by image and no.of.pixels that each class has.   

I have tried following steps to overcome this class imbalance:

**1.I have calculated the class weights based on the procedure mentioned here** https://github.com/qubvel/segmentation_models/issues/137#issuecomment-515774767 

And obtained my class weights as [  1.52601775  27.67985103 126.29271733 402.83724388  72.80682642  86.53852883   6.85154221  ]

since  class weights are recommended to be [0-1]. I converted the above weights to decimal [0.0037 0.068 0.3141 1.002  0.1811 0.2152 0.017 ]

**2. And I tried with combined focal and dice loss with and without class weights.**

**3. Also,I tried with both encoder_freeze = True** and **encoder_freeze = False**

**4.I have also checked that my masks are in the range[0-1]**

I am using UNET with resnet-50 backbone. 

But the val_loss does not decrease,and it continues to fluctuate above 93% and also I am not able to get good IOU score.

Please give me some inputs to overcome the above issue and also i need to know how to print the class wise metrics

Thanks in advance

**Comment by JordanMakesMaps** [comment]
@Indhujamari just throwing out extra things for you to try, but have you tried:

- using the class_weight parameter in model.fit() instead of using with the loss function(s)? Don't rescale from [0, 1] with this one;
- undersampling majority/over sampling minority classes? For example, if every image has class A, but only a few images have class B - G, then remove all images that only contain class A.
- using categorical_cross_entropy + jaccard
- using one of the efficientnet as the encoder, w/ imagenet or nosy-student weights?

**Comment by Indhujamari** [comment]
@JordanMakesMaps Thank you for your inputs. Let me try it out.

**Comment by nivetha1234** [comment]
I also have the same issue with multi class segmentation where my dataset contains about 8 classes (6 classes + 1 background + 1 ignore) which are highly imbalanced.But how to manually assign class weights to this imbalanced dataset... Please help me to solve this issue. Thanks in advance!

**Comment by nivetha1234** [comment]
@JordanMakesMaps  Thank you for your solution! I'll definitely try it.

**Comment by murdav** [comment]
@nivetha1234, did you do something in particular to update the model? E.g which models, loss and metrics did you use?
 Thank you.
